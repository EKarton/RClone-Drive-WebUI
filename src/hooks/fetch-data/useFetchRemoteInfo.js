import { useCallback } from 'react';
import useFetchData from './useFetchData';

export default function useFetchRemoteInfo(remote) {
  const func = (c, cancelToken) => c.fetchRemoteInfo(remote, { cancelToken });
  return useFetchData(useCallback(func, [remote]));
}
