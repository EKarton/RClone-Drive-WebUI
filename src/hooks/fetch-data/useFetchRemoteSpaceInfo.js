import { useCallback } from 'react';
import useFetchData from './useFetchData';

export default function useFetchRemoteSpaceInfo(remote) {
  const func = (c, cancelToken) => c.fetchRemoteSpaceInfo(remote, { cancelToken });
  return useFetchData(useCallback(func, [remote]));
}
