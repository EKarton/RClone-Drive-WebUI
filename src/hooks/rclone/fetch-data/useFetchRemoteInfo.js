import useFetchData from './useFetchData';
import { useCallback } from 'react';

export default function useFetchRemoteInfo(remote) {
  const func = (c, cancelToken) => c.fetchRemoteInfo(remote, { cancelToken });
  return useFetchData(useCallback(func, [remote]));
}
