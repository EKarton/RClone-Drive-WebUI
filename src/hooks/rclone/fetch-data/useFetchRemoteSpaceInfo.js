import useFetchData from './useFetchData';
import { useCallback } from 'react';

export default function useFetchRemoteSpaceInfo(remote) {
  const func = (c, cancelToken) => c.fetchRemoteSpaceInfo(remote, { cancelToken });
  return useFetchData(useCallback(func, [remote]));
}
