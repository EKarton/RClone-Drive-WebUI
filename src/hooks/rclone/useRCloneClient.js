import RCloneClient from 'services/RCloneClient';
import useRCloneInfo from './useRCloneInfo';

let rCloneInstance = null;

/**
 * Returns the singleton of the RCloneClient
 * If the singleton does not exist yet, it will create one
 *
 * @returns {RCloneClient} the RClone client
 */
export default function useRCloneClient() {
  const { rCloneInfo } = useRCloneInfo();

  if (!rCloneInfo.endpoint || !rCloneInfo.username || !rCloneInfo.password) {
    throw new MissingRCloneInfoError();
  }

  if (rCloneInstance) {
    return rCloneInstance;
  }

  rCloneInstance = new RCloneClient(rCloneInfo, rCloneInfo, rCloneInfo);

  return rCloneInstance;
}

export class MissingRCloneInfoError extends Error {
  constructor() {
    super('Missing RClone info!');
  }
}
