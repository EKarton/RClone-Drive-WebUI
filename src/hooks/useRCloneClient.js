import { useMemo } from "react";
import RCloneClient from "utils/RCloneClient";
import useRCloneInfo from "./useRCloneInfo";

export default function useRCloneClient() {
  const { rCloneInfo } = useRCloneInfo();

  if (!rCloneInfo.endpoint || !rCloneInfo.username || !rCloneInfo.password) {
    throw new MissingRCloneInfoError();
  }

  const client = useMemo(() => {
    return new RCloneClient(
      rCloneInfo.endpoint,
      rCloneInfo.username,
      rCloneInfo.password
    );
  }, [rCloneInfo.endpoint, rCloneInfo.username, rCloneInfo.password]);

  return client;
}

export class MissingRCloneInfoError extends Error {
  constructor() {
    super("Missing RClone info!");
  }
}
