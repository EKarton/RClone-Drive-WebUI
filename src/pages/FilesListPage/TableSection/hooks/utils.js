import { JobStatus, JobTypes } from 'utils/constants';

export function getFileNames(data) {
  return new Set(data.filter((file) => !file.IsDir).map((file) => file.Name));
}

export function getFolderNames(data) {
  return new Set(data.filter((file) => file.IsDir).map((file) => file.Name));
}

export function getUploadingJobs(jobs, remote, path) {
  return jobs.filter(
    (job) =>
      job.jobType === JobTypes.UPLOAD_FILE &&
      job.remote === remote &&
      job.dirPath.startsWith(path) &&
      job.status.value !== JobStatus.SUCCESS
  );
}

export function getUploadingFiles(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath === path);
}

export function getUploadingFolders(jobs, remote, path) {
  return getUploadingJobs(jobs, remote, path).filter((job) => job.dirPath !== path);
}

export function getMovingJobs(jobs, remote, path) {
  const supportedJobTypes = new Set([
    JobTypes.MOVE_FILE,
    JobTypes.MOVE_FOLDER,
    JobTypes.RENAME_FILE,
    JobTypes.RENAME_FOLDER,
  ]);

  return jobs.filter((job) => {
    const correctJob = supportedJobTypes.has(job.jobType);
    const inProgress = job.status.value === JobStatus.ONGOING;

    if (!correctJob || !inProgress) {
      return false;
    }

    const isMatchSrc = job.src?.remote === remote && job.src?.dirPath === path;
    const isMatchTarget = job.target?.remote === remote && job.target?.dirPath === path;

    return correctJob && inProgress && (isMatchSrc || isMatchTarget);
  });
}
