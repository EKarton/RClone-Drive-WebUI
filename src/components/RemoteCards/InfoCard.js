import DeleteSweepIcon from '@mui/icons-material/DeleteSweep';
import ListItemIcon from '@mui/material/ListItemIcon';
import ListItemText from '@mui/material/ListItemText';
import Menu from '@mui/material/Menu';
import MenuItem from '@mui/material/MenuItem';
import Skeleton from '@mui/material/Skeleton';
import prettyBytes from 'pretty-bytes';
import PropTypes from 'prop-types';
import { useState } from 'react';
import useFetchRemoteInfo from 'hooks/fetch-data/useFetchRemoteInfo';
import useFetchRemoteSpaceInfo from 'hooks/fetch-data/useFetchRemoteSpaceInfo';
import useRCloneClient from 'hooks/rclone/useRCloneClient';
import { StatusTypes } from 'utils/constants';
import BaseCard from './BaseCard';

export default function InfoCard({ remote, onClick }) {
  const sizeResult = useFetchRemoteSpaceInfo(remote);
  const infoResult = useFetchRemoteInfo(remote);

  const rCloneClient = useRCloneClient();
  const [isClearingTrashCan, setIsClearingTrashCan] = useState(false);
  const [contextMenuPos, setContextMenuPos] = useState(undefined);
  const [isContextMenuOpen, setIsContextMenuOpen] = useState(false);

  const renderRemoteInfo = () => {
    if (infoResult.status === StatusTypes.LOADING || isClearingTrashCan) {
      return <Skeleton data-testid="remote-info-skeleton" />;
    }

    if (infoResult.status === StatusTypes.ERROR) {
      return 'Unable to get remote details';
    }

    return `Type: ${infoResult.data?.type}`;
  };

  const renderSpaceUsed = () => {
    if (sizeResult.status === StatusTypes.LOADING || isClearingTrashCan) {
      return <Skeleton data-testid="remote-space-skeleton" />;
    }

    if (sizeResult.status === StatusTypes.ERROR) {
      return 'Unable to get space information';
    }

    const { total, used, trashed } = sizeResult.data;
    const totalSpace = Number.isInteger(total) ? prettyBytes(total) : 'NA';
    const spaceUsed = Number.isInteger(used) ? prettyBytes(used) : 'NA';
    const trashUsed = Number.isInteger(trashed) ? prettyBytes(trashed) : 'NA';

    return `${spaceUsed} / ${totalSpace} used, ${trashUsed} in trash`;
  };

  const handleContextMenu = (e) => {
    e.preventDefault();

    setContextMenuPos({ left: e.clientX - 2, top: e.clientY - 4 });
    setIsContextMenuOpen(true);
  };

  const handleClearTrashClick = async () => {
    try {
      setIsClearingTrashCan(true);
      handleCloseContextMenu();

      await rCloneClient.emptyTrashCan(remote);
    } catch (err) {}

    setIsClearingTrashCan(false);
    sizeResult.refetchData();
    infoResult.refetchData();
  };

  const handleCloseContextMenu = () => {
    setIsContextMenuOpen(false);
    setContextMenuPos(undefined);
  };

  return (
    <>
      <BaseCard
        remoteName={remote}
        remoteType={renderRemoteInfo()}
        remoteSpace={renderSpaceUsed()}
        onClick={onClick}
        onContextMenu={handleContextMenu}
      />
      <Menu
        open={isContextMenuOpen}
        onClose={handleCloseContextMenu}
        anchorReference="anchorPosition"
        anchorPosition={contextMenuPos}
      >
        <MenuItem onClick={handleClearTrashClick} data-testid="clear-trash-button">
          <ListItemIcon>
            <DeleteSweepIcon fontSize="small" />
          </ListItemIcon>
          <ListItemText>Clear Trash</ListItemText>
        </MenuItem>
      </Menu>
    </>
  );
}

InfoCard.propTypes = {
  remote: PropTypes.string,
  onClick: PropTypes.func,
};
