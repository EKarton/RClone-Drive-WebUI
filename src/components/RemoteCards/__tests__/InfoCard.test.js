import useFetchRemoteInfo from 'hooks/fetch-data/useFetchRemoteInfo';
import useFetchRemoteSpaceInfo from 'hooks/fetch-data/useFetchRemoteSpaceInfo';
import useRCloneClient from 'hooks/rclone/useRCloneClient';
import { StatusTypes } from 'utils/constants';
import { mockOperationsAboutResponse } from 'test-utils/mock-responses';
import { mockConfigGetResponse } from 'test-utils/mock-responses';
import { customRender, screen, fireEvent, userEvent } from 'test-utils/react';
import InfoCard from '../InfoCard';

jest.mock('hooks/fetch-data/useFetchRemoteSpaceInfo');
jest.mock('hooks/fetch-data/useFetchRemoteInfo');
jest.mock('hooks/rclone/useRCloneClient');

describe('InfoCard', () => {
  const emptyTrashCan = jest.fn();

  beforeEach(() => {
    emptyTrashCan.mockResolvedValue();

    useRCloneClient.mockReturnValue({
      emptyTrashCan,
    });

    useFetchRemoteSpaceInfo.mockReturnValue({
      status: StatusTypes.SUCCESS,
      data: mockOperationsAboutResponse,
      refetchData: jest.fn(),
    });

    useFetchRemoteInfo.mockReturnValue({
      status: StatusTypes.SUCCESS,
      data: mockConfigGetResponse,
      refetchData: jest.fn(),
    });
  });

  it('should render spinners when api call is in flight', async () => {
    useFetchRemoteSpaceInfo.mockReturnValue({ status: StatusTypes.LOADING });
    useFetchRemoteInfo.mockReturnValue({ status: StatusTypes.LOADING });

    const { baseElement } = customRender(<InfoCard remote="googledrive" />);

    await screen.findByTestId('remote-info-skeleton');
    await screen.findByTestId('remote-space-skeleton');
    expect(baseElement).toMatchSnapshot();
  });

  it('should render data correctly when api call finishes', async () => {
    const { baseElement } = customRender(<InfoCard remote="googledrive" />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should render data correctly given there is no data in its size api call', async () => {
    useFetchRemoteSpaceInfo.mockReturnValue({
      status: StatusTypes.SUCCESS,
      data: {
        total: undefined,
        used: undefined,
        trashed: undefined,
      },
      refetchData: jest.fn(),
    });

    const { baseElement } = customRender(<InfoCard remote="googledrive" />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should render error messages correctly when api calls fail', async () => {
    useFetchRemoteSpaceInfo.mockReturnValue({
      status: StatusTypes.ERROR,
      error: new Error('404 not found'),
    });
    useFetchRemoteInfo.mockReturnValue({
      status: StatusTypes.ERROR,
      error: new Error('404 not found'),
    });

    const { baseElement } = customRender(<InfoCard remote="googledrive" />);

    await screen.findByText('Unable to get remote details');
    await screen.findByText('Unable to get space information');
    expect(baseElement).toMatchSnapshot();
  });

  it('should call RCloneClient.emptyTrashCan() when user right-clicks on card and selects Clear Trash', () => {
    customRender(<InfoCard remote="googledrive" />);

    fireEvent.contextMenu(screen.getByRole('button'));
    userEvent.click(screen.getByTestId('clear-trash-button'));

    expect(emptyTrashCan).toBeCalledWith('googledrive');
  });
});
