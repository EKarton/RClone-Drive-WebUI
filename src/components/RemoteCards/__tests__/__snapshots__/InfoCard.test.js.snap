// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoCard should render data correctly given there is no data in its size api call 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
          >
            googledrive
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Type: crypt
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            NA / NA used, NA in trash
          </p>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`InfoCard should render data correctly when api call finishes 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
          >
            googledrive
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Type: crypt
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            126 GB / 161 GB used, 0 B in trash
          </p>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`InfoCard should render error messages correctly when api calls fail 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
          >
            googledrive
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Unable to get remote details
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            Unable to get space information
          </p>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`InfoCard should render spinners when api call is in flight 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
    >
      <button
        class="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h6 remote-base-card__name css-2ulfj5-MuiTypography-root"
          >
            googledrive
          </div>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              data-testid="remote-info-skeleton"
            />
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1n7gpw1-MuiSkeleton-root"
              data-testid="remote-space-skeleton"
            />
          </p>
        </div>
        <span
          class="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
        />
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;
